### 1. Problem & Project Definition

i. Business problem  
A real estate company wants to understand   property value factors and support better pricing decisions.

ii. Decision lens 
→ Price Prediction 

iii.  Stakeholder and Decision need
| No. | Stakeholder | Decision Need |
|---|---|---|
||**Primary Stakeholders**|| 
| 1 | **Real Estate Agents & Brokers** | To estimate a property's market value from its characteristics and use the estimate to set a realistic and competitive listing price. |
|| **Secondary Stakeholders**||  
| 1 | **Certified Appraisers** | To use the estimated property value as supporting information when deciding an appropriate asking price for the property. |
| 2 | **The Seller (Property Owner)** | To use data-driven property-value estimates as supplementary information when assessing the value of a property. |
| 3 | **Property Investors & Developers** | To use estimated property values to support property purchasing and investment decisions. |
| 4 | **Buyers** | To check if the asking price is fair so they do not overpay for a home. |
     
iv. Unit of analysis 
→ property

v. ML task 
→ Regression
Because here we predict the sales price not just choosing a class

vi. Target 
→ Sale Price

vii. Expected output 
→ estimated Sale Price

viii. Why ML is suitable
ML is suitable for this problem because it can learn the relationships between different property features and their sale prices from historical data. The trained model can then use these patterns to predict the sale price of a new property and support better pricing decisions.


### 2. Dataset Understanding

In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import gdown

##### 2.1 Load Training Dataset

In [19]:
file_id = "1To_n4yys1lnWwbQ7bvRDYW8B_f2gfEud"

gdown.download(
    id=file_id,
    output="dataset.csv",
    quiet=False
)

df = pd.read_csv("dataset.csv")

df.head()

Downloading...
From: https://drive.google.com/uc?id=1To_n4yys1lnWwbQ7bvRDYW8B_f2gfEud
To: c:\1. SLIIT\Y3S1\ML\Assignment\ames-housing-price-prediction\dataset.csv
100%|██████████| 461k/461k [00:00<00:00, 1.12MB/s]


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [20]:
# df = pd.read_csv("https://drive.google.com/file/d/1Ys_QhDgt_15kFixVB6SwSv24afhKiOQf/view?usp=drive_link")

##### 2.2 Dataset Dimensions

In [21]:
df.shape

(1460, 81)

##### 2.3 Initial Dataset Inspection

In [22]:
print('First 5 rows of the DataFrame:\n', df.head())

First 5 rows of the DataFrame:
    Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  S

In [23]:
print('Additional information:\n ')
df.info()

Additional information:
 
<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   str    
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   str    
 6   Alley          91 non-null     str    
 7   LotShape       1460 non-null   str    
 8   LandContour    1460 non-null   str    
 9   Utilities      1460 non-null   str    
 10  LotConfig      1460 non-null   str    
 11  LandSlope      1460 non-null   str    
 12  Neighborhood   1460 non-null   str    
 13  Condition1     1460 non-null   str    
 14  Condition2     1460 non-null   str    
 15  BldgType       1460 non-null   str    
 16  HouseStyle     1460 non-null   str    
 17  OverallQual    1460 non-null   int64 

##### 2.4 Features and Target

In [24]:
df.columns.tolist()

['Id',
 'MSSubClass',
 'MSZoning',
 'LotFrontage',
 'LotArea',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'MasVnrArea',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinSF1',
 'BsmtFinType2',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'KitchenQual',
 'TotRmsAbvGrd',
 'Functional',
 'Fireplaces',
 'FireplaceQu',
 'GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageCars',
 'GarageArea',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'WoodDeckSF',
 'OpenPorchSF'

Target

In [25]:
df['SalePrice'].dtype

dtype('int64')

In [26]:
df["SalePrice"].head(3)

0    208500
1    181500
2    223500
Name: SalePrice, dtype: int64

In [27]:
df["SalePrice"].describe()

count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

##### 2.5 Data Types

In [28]:
pd.set_option('display.max_rows', None)
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning             str
LotFrontage      float64
LotArea            int64
Street               str
Alley                str
LotShape             str
LandContour          str
Utilities            str
LotConfig            str
LandSlope            str
Neighborhood         str
Condition1           str
Condition2           str
BldgType             str
HouseStyle           str
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle            str
RoofMatl             str
Exterior1st          str
Exterior2nd          str
MasVnrType           str
MasVnrArea       float64
ExterQual            str
ExterCond            str
Foundation           str
BsmtQual             str
BsmtCond             str
BsmtExposure         str
BsmtFinType1         str
BsmtFinSF1         int64
BsmtFinType2         str
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating              str


Binary Variables: Utilities, Street and CentralAir are binary categorical variables because each contains two possible categories.

##### 2.6 Numerical Variables

In [29]:
numerical_columns = df.select_dtypes(include=['int64', 'float64']).columns

numerical_columns.tolist()

['Id',
 'MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold',
 'SalePrice']

In [30]:
len(numerical_columns)

38

In [31]:
df[numerical_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.500000,421.610009,1.0,365.75,730.5,1095.25,1460.0
MSSubClass,1460.0,56.897260,42.300571,20.0,20.00,50.0,70.00,190.0
LotFrontage,1201.0,70.049958,24.284752,21.0,59.00,69.0,80.00,313.0
LotArea,1460.0,10516.828082,9981.264932,1300.0,7553.50,9478.5,11601.50,215245.0
OverallQual,1460.0,6.099315,1.382997,1.0,5.00,6.0,7.00,10.0
OverallCond,1460.0,5.575342,1.112799,1.0,5.00,5.0,6.00,9.0
YearBuilt,1460.0,1971.267808,30.202904,1872.0,1954.00,1973.0,2000.00,2010.0
YearRemodAdd,1460.0,1984.865753,20.645407,1950.0,1967.00,1994.0,2004.00,2010.0
MasVnrArea,1452.0,103.685262,181.066207,0.0,0.00,0.0,166.00,1600.0
BsmtFinSF1,1460.0,443.639726,456.098091,0.0,0.00,383.5,712.25,5644.0


##### 2.7 Categorical Variables

In [32]:
categorical_columns = df.select_dtypes(include=['object']).columns

categorical_columns.tolist()

C:\Users\dampa\AppData\Local\Temp\ipykernel_3700\2750784672.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include=['object']).columns


['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [34]:
len(categorical_columns)

43

##### Looking at the types of each category

In [35]:
for column in categorical_columns:
    print(f"\n{column}")
    print(df[column].unique())


MSZoning
<StringArray>
['RL', 'RM', 'C (all)', 'FV', 'RH']
Length: 5, dtype: str

Street
<StringArray>
['Pave', 'Grvl']
Length: 2, dtype: str

Alley
<StringArray>
[nan, 'Grvl', 'Pave']
Length: 3, dtype: str

LotShape
<StringArray>
['Reg', 'IR1', 'IR2', 'IR3']
Length: 4, dtype: str

LandContour
<StringArray>
['Lvl', 'Bnk', 'Low', 'HLS']
Length: 4, dtype: str

Utilities
<StringArray>
['AllPub', 'NoSeWa']
Length: 2, dtype: str

LotConfig
<StringArray>
['Inside', 'FR2', 'Corner', 'CulDSac', 'FR3']
Length: 5, dtype: str

LandSlope
<StringArray>
['Gtl', 'Mod', 'Sev']
Length: 3, dtype: str

Neighborhood
<StringArray>
['CollgCr', 'Veenker', 'Crawfor', 'NoRidge', 'Mitchel', 'Somerst',  'NWAmes',
 'OldTown', 'BrkSide',  'Sawyer', 'NridgHt',   'NAmes', 'SawyerW',  'IDOTRR',
 'MeadowV', 'Edwards',  'Timber', 'Gilbert', 'StoneBr', 'ClearCr', 'NPkVill',
 'Blmngtn',  'BrDale',   'SWISU', 'Blueste']
Length: 25, dtype: str

Condition1
<StringArray>
['Norm', 'Feedr', 'PosN', 'Artery', 'RRAe', 'RRNn', '

##### 2.8 Ordinal Variables

In [36]:
ordinal_columns = [
    'LotShape',
    'LandSlope',
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'BsmtExposure',
    'BsmtFinType1',
    'HeatingQC',
    'KitchenQual',
    'Functional',
    'FireplaceQu',
    'GarageFinish',
    'GarageQual',
    'GarageCond',
    'PavedDrive',
    'PoolQC',
    'Fence'

]

_Ordinal variables are categories that have a meaningful order or ranking.  In this dataset, these are the several variables that represent such level. The identified ordinal variables and their number of unique categories are shown below._

In [37]:
df[ordinal_columns].nunique().sort_values()

LandSlope       3
PavedDrive      3
GarageFinish    3
PoolQC          3
BsmtExposure    4
BsmtCond        4
ExterQual       4
LotShape        4
KitchenQual     4
BsmtQual        4
Fence           4
ExterCond       5
GarageCond      5
FireplaceQu     5
GarageQual      5
HeatingQC       5
BsmtFinType1    6
Functional      7
dtype: int64

##### 2.9 Understanding Variables

1. House / Property Characteristics
- LotFrontage — Linear feet of street connected to property
- LotArea — Lot size
- LotShape — Shape of property
- LandContour — Flatness/contour
- LandSlope — Slope
- LotConfig — Lot configuration
- BldgType — Type of dwelling
- HouseStyle — Style of dwelling
- OverallQual — Overall material/finish quality
- OverallCond — Overall condition
- YearBuilt — Construction year
- YearRemodAdd — Remodeling year
- 1stFlrSF — First-floor area
- 2ndFlrSF — Second-floor area
- GrLivArea — Above-ground living area
- BedroomAbvGr / Bedroom — Bedrooms
- KitchenAbvGr / Kitchen — Kitchens
- TotRmsAbvGrd — Total rooms
- FullBath — Full bathrooms
- HalfBath — Half bathrooms
- BsmtFullBath — Basement full bathrooms
- BsmtHalfBath — Basement half bathrooms

2. Basement Features
- BsmtQual --> height of the basement
- BsmtCond --> general condition of the basement
- BsmtExposure --> walkout or garden level walls
- BsmtFinType1 --> Rating of basement finished area
- BsmtFinType2 --> Rating of basement finished are
- BsmtFinSF1 --> Type 1 finished square feet
- BsmtFinSF2 --> Type 2 finished square feet
- BsmtUnfSF --> Unfinished square feet of basement area
- TotalBsmtSF --> Total square feet of basement area

3. Garage Features
- GarageType --> Garage location
- GarageYrBlt --> Year garage was built
- GarageFinish --> Interior finish of the garage
- GarageCars --> Size of garage in car capacity
- GarageArea --> Size of garage in square feet
- GarageQual --> Garage quality
- GarageCond --> Garage condition

4. House Systems / Utilities
- Heating --> Type of heating
- HeatingQC --> Heating quality and condition
- CentralAir --> Central air conditioning
- Electrical --> Electrical system
- Utilities --> Type of utilities available

5. Exterior / Construction
- RoofStyle --> Type of roof
- RoofMatl --> Roof material
- Exterior1st --> Exterior covering on house
- Exterior2nd --> Exterior covering on house (if more than one material)
- MasVnrType --> Masonry veneer type
- MasVnrArea --> Masonry veneer area in square feet
- ExterQual --> quality of the material on the exterior
- ExterCond --> present condition of the material on the exterior
- Foundation --> Type of foundation

6. Outdoor Features
- WoodDeckSF --> Wood deck area in square feet
- OpenPorchSF --> Open porch area in square feet
- EnclosedPorch --> Enclosed porch area in square feet
- 3SsnPorch --> Three season porch area in square feet
- ScreenPorch --> Screen porch area in square feet
- PoolArea --> Pool area in square feet
- PoolQC --> Pool quality
- Fence --> Fence quality
- MiscFeature --> Miscellaneous feature not covered in other categories
- MiscVal --> Value of miscellaneous feature

7. Location / Access
- MSZoning --> general zoning classification of the sale
- Street --> Type of road access to property
- Alley --> Type of alley access to property
- Neighborhood --> Physical locations within Ames city limits
- Condition1 --> Proximity to various conditions
- Condition2 --> Proximity to various conditions (if more than 1 is present)


8. Sale / Transaction Features
- MoSold
- YrSold
- SaleType
- SaleCondition

The dataset contains 81 columns and 1,460 observations. The target variable is SalePrice. Among the 81 columns, 38 are numerical variables and 43 are categorical variables. Since SalePrice is the numerical target variable, there are 37 numerical predictor features and 43 categorical predictor features, giving a total of 80 input features.

By examining the categorical variables, we also identified categorical features with a meaningful order, which can be considered ordinal variables. In addition, two categorical variables, Street and CentralAir, are binary variables because each contains two possible categories. These represent the main data types identified in the dataset.